File: Data-Structures/linkedlist/cLinkedList.c
#include <stdio.h>
#include <stdlib.h>

struct node {
    int data;
    struct node *next;
};

void fnDisplay(struct node *head){
    struct node *cur;
    int count=0;
    if(head==NULL){
        printf("List is empty\n");
        return;
    }
    cur=head;
    printf("Elements of linked list are : ");
    do{
        printf("%d\t",cur->data);
        cur=cur->next;
        count++;
    }while(cur!=head);
    printf("\nTotal no of elements are : %d\n",count);
}

void insertAtEnd(struct node **head,int data){
    struct node *newNode,*cur;
    newNode=(struct node *)malloc(sizeof(struct node));
    if(!newNode){
        printf("Memory Error\n");
        return;
    }
    newNode->data=data;
    newNode->next=newNode;
    if(*head){  
        cur=*head;
        while(cur->next!=*head)
                cur=cur->next;
        cur->next=newNode;
        newNode->next=*head;
    }
    else
        *head=newNode;
}

void insertAtFront(struct node **head,int data){
    struct node *cur,*newNode;
    newNode=(struct node *)malloc(sizeof(struct node));
    if(!newNode){
        printf("Memory error\n");
        return;
    }
    newNode->data=data;
    newNode->next=newNode;
    if(*head){
        cur=*head;
        while(cur->next!=*head)
            cur=cur->next;
        newNode->next=*head;
        cur->next=newNode;
    }
        *head=newNode;
}

void deleteLastNode(struct node **head){
    struct node *cur,*prev;
    if(*head==NULL){
        printf("List is empty\n");
        return;
    }
    if(*head){
        prev=*head;
        cur=*head;
        while(cur->next!=*head){
            prev=cur;
            cur=cur->next;
        }
        prev->next=*head;
        free(cur);
    }
}

void deleteFirstNode(struct node **head){
    struct node *first,*cur;
    if(*head==NULL){
        printf("List is empty\n");
        return;
    }
    if(*head){
        first=*head;
        cur=*head;
        while(cur->next!=*head)
            cur=cur->next;
        cur->next=first->next;
        *head=first->next;
        free(first);
    }
}

int main()
{
    int data,choice;
    struct node *head=malloc(sizeof(struct node));
    if(!head){
        printf("Memory error \n");
        return 0;
    }
    head=NULL;
    while(1){
        printf("Choose operation : 1.Insert at end 2.Insert at front 3.Delete at end 4.Delete at front 5.Display\n");
        scanf("%d",&choice);
        switch (choice)
        {
        case 1: printf("Enter data : ");
                scanf("%d",&data);
                insertAtEnd(&head,data);
                break;
        case 2: printf("Enter data : ");
                scanf("%d",&data);
                insertAtFront(&head,data);
                break;
        case 3: deleteLastNode(&head);
                break;
        case 4: deleteFirstNode(&head);
                break;
        case 5: fnDisplay(head);
                break;
        default: printf("Invalid input\n");
                 break;
        }
    }
    return 0;
}

File: Data-Structures/linkedlist/cLinkedList.java
import java.util.Scanner;

class node{
    int data;
    node next;
    node(int data){
        this.data=data;
    }
}

class cLinkedList{
    public static node head;
    static void fnDisplay(){
        node cur;
        int count=0;
        if(head==null){
            System.out.println("List is empty");
            return;
        }
        cur=head;
        System.out.println("Elements of Circular Linked List are : ");
        do{
            System.out.println(cur.data);
            cur=cur.next;
            count++;
        }while(cur!=head);
        System.out.println("Total no of elements of List are : " + count);
    }

    static void fnInsertAtEnd(int data){
        node cur,newNode;
        newNode = new node(data);
        newNode.next=newNode;
        if(head==null){
            head=newNode;
            return;
        }
        cur=head;
        do{
            cur=cur.next;
        }while(cur.next!=head);
        cur.next=newNode;
        newNode.next=head;
    }

    static void fninsertAtFront(int data){
        node first,cur,newNode;
        newNode = new node(data);
        newNode.next=newNode;
        if(head==null){
            head=newNode;
            return;
        }
        first=head;
        cur=head;
        do{
            cur=cur.next;
        }while(cur.next!=head);
        cur.next=newNode;
        newNode.next=first;
        head=newNode;
    }

    static void fnDeleteLastNode(){
        node prev,cur;
        if(head==null){
            System.out.println("List is empty");
            return;
        }
        prev=head;
        cur=head;
        do{
            prev=cur;
            cur=cur.next;
        }while(cur.next!=head);
        if(prev==cur){
            head=null;
        }
        else{
            prev.next=head;
        }
    }

    static void fnDeleteFirstNode(){
        node cur;
        if(head==null){
            System.out.println("List is empty");
            return;
        }
        cur=head;
        do{
            cur=cur.next;
        }while(cur.next!=head);
        if(head.next!=head){
            head=head.next;
            cur.next=head;
        }
        else{
            head=null;
        }
    }

    public static void main(String[] args) {
        int data,choice;
        Scanner scanner = new Scanner(System.in);
        while(true){
            System.out.println("Select operation : 1.Insert at front 2.Insert at rear 3.Delete at front 4.Delete at rear 5.Display 6.Exit");
            choice = scanner.nextInt();
            switch(choice){
                case 1: System.out.println("Enter data");
                        data=scanner.nextInt();
                        fninsertAtFront(data);
                        break;
                case 2: System.out.println("Enter data");
                        data=scanner.nextInt();
                        fnInsertAtEnd(data);
                        break;
                case 3: fnDeleteFirstNode();
                        break;
                case 4: fnDeleteLastNode();
                        break;
                case 5: fnDisplay();
                        break;
                case 6: System.exit(1);
                default:System.out.println("Invalid input");
                        break;

            }
        }
    }
}